Sources/PackageLoading/PackageBuilder.swift

```diff
@@ -234,54 +244,67 @@ public struct PackageBuilder {
             let sources = Sources(paths: [moduleMapPath], root: packagePath)
             return [try CModule(name: manifest.name, sources: sources, path: packagePath, pkgConfig: pkgConfigPath, providers: manifest.package.providers)]
         }
 
+        // If everything is excluded, just return an empty array.
         if manifest.package.exclude.contains(".") {
             return []
         }
-
-        let srcroot = try sourceRoot()
-
-        if srcroot != packagePath {
+        
+        // Locate the source directory inside the package.
+        let srcDir = try sourceRoot()
+        
+        // If there is a source directory, we expect all source files to be located in it.
+        if srcDir != packagePath {
             let invalidRootFiles = try directoryContents(packagePath).filter(isValidSource)
             guard invalidRootFiles.isEmpty else {
                 throw ModuleError.invalidLayout(.invalidLayout(invalidRootFiles.map{ $0.asString }))
             }
         }
-
-        let maybeModules = try directoryContents(srcroot).filter(shouldConsiderDirectory)
-
-        if maybeModules.count == 1 && maybeModules[0] != srcroot {
-            let invalidModuleFiles = try directoryContents(srcroot).filter(isValidSource)
+        
+        // Locate any directories that might be the roots of modules inside the source directory.
+        let potentialModulePaths = try directoryContents(srcDir).filter(shouldConsiderDirectory)
+        
+        // If there's a single module inside the source directory, make sure there are no loose source files in the sources directory.
+        if potentialModulePaths.count == 1 && potentialModulePaths[0] != srcDir {
+            let invalidModuleFiles = try directoryContents(srcDir).filter(isValidSource)
             guard invalidModuleFiles.isEmpty else {
                 throw ModuleError.invalidLayout(.invalidLayout(invalidModuleFiles.map{ $0.asString }))
             }
         }
-
+        
+        // With preliminary checks done, we can start creating modules.
         let modules: [Module]
-        if maybeModules.isEmpty {
-            // If there are no sources subdirectories, we have at most a one target package.
+        if potentialModulePaths.isEmpty {
+            // There are no directories that look like modules, so try to create a module for the source directory itself (with the name coming from the name in the manifest).
             do {
-                modules = [try modulify(srcroot, name: manifest.name, isTest: false)]
-            } catch Module.Error.noSources {
+                modules = [try createModule(srcDir, name: manifest.name, isTest: false)]
+            }
+            catch Module.Error.noSources {
                 // Completely empty packages are allowed as a special case.
                 modules = []
             }
-        } else {
-            modules = try maybeModules.map { path in
+        }
+        else {
```

Comment:
This is a nit, but can we use `} else {` which is most consistent with the rest of the code and Swift?